Add readme with guide for package authors (#3567)
JuliaRegistries/General: docs/src/guide.md

# Guide for package authors

New packages, or a new version of packages, are added the Julia General registry by pull requests against the [JuliaRegistries/General](https://github.com/JuliaRegistries/General/pulls) GitHub repository. This is automated by the [Registrator.jl](https://github.com/JuliaRegistries/Registrator.jl) GitHub bot and the [web interface](https://github.com/JuliaRegistries/Registrator.jl). 

It can be helpful to also install [TagBot](https://github.com/JuliaRegistries/TagBot), which automatically tags a release on your repository after the package or new version of the package is added to the registry.

The process usually proceeds as follows:

> Either

Reviewer: ```suggestion
Either:

* Install the Registrator bot on the repository corresponding to the package you would like to add or update in the General registry, following the instructions in the readme at [Registrator.jl](https://github.com/JuliaRegistries/Registrator.jl). Then file an issue on the repository, or make a comment on the commit, with "@JuliaRegistrator register". For more information on interacting with the Registrator bot, see the aforementioned README.
* Go to the [web interface](https://github.com/JuliaRegistries/Registrator.jl) and follow the instructions linked there to start the registration process.

After either option, a pull request is made against the [JuliaRegistries/General](https://github.com/JuliaRegistries/General/pulls) GitHub repository.

> Often, that is all one needs to do. New packages are subject to a 3-day waiting period for community comment, while new versions of existing packages may be merged right away. Once the pull request is merged, the package or new version of the package is available in the General registry. If TagBog is installed, then it will tag a release once the pull request is merged. This is very helpful to ensure that tagged releases correspond exactly to versions in the registry.

Reviewer: ```suggestion
Often, that is all one needs to do. New packages are subject to a 3-day waiting period for community comment, while new versions of existing packages may be merged right away. Once the pull request is merged, the package or new version of the package is available in the General registry. If [TagBot](https://github.com/JuliaRegistries/TagBot) is installed, then it will tag a release once the pull request is merged. This is very helpful to ensure that tagged releases correspond exactly to versions in the registry.

Before the pull request is merged, retriggering the Registrator bot on the same package (via a new comment or use of the web interface) will cause the bot to update the pull request (by force pushing).

While usually starting the registration process is all the package author needs to do, occasionally, however, something more may be needed. In this case, the "waiting for author" label will be applied to the pull request, and a maintainer will comment on the pull request. Several common scenarios and what to do are listed below; these are meant to be linked to help explain a particular situation when it arises.

## Compatability ("needs compat")

> Every Julia package has at least one (implicit) dependency: the Julia language itself. It is recommended that for every dependency, each package declares what versions of the dependency it is compatible with. This is done in the package's `Project.toml` file, and is documented [here](https://julialang.github.io/Pkg.jl/v1/compatibility/) in the `Pkg` standard library's doucmentation.

Reviewer: ```suggestion
Every Julia package has at least one (implicit) dependency: the Julia language itself. It is recommended that for every dependency, each package declares what versions of the dependency it is compatible with. This is done in the package's `Project.toml` file, and is documented [here](https://julialang.github.io/Pkg.jl/v1/compatibility/) in the Julia package manager's documentation.
```

Might not be necessary, I felt it made it clearer though.

Packages should at least declare the Julia version they are compatible with, by e.g.

```toml
[compat]
julia = "1"
```

which declares that the package is compatible with all Julia versions from 1.0 onwards, until but not including, a possible 2.0 in the future.

Tt is recommended to have one entry in the `compat` section of `Project.toml` for every dependency, not just Julia itself. Note that standard libraries are currently coupled to the Julia version, and don't have separate version numbers. So while they are declared as dependencies, they don't need separate entries in the `compat` section from Julia itself.

## Merge conflicts (i.e. "needs rebase")

If the pull request against the General registry has merge conflicts with the General registry itself, the maintainers of the General registry will be unable to merge the pull request. In this case, they will add the "waiting for author" tag and comment that there are merge conflicts or that the pull request needs a rebase. Just retrigger the Registrator.jl bot (either add another comment with "@JuliaRegistrator register", or use the web interface), and it will update the pull request.

## Julia Package naming conventions

> The guidelines for naming packages are listed [here](https://julialang.github.io/Pkg.jl/v1/creating-packages/#Package-naming-guidelines-1) in the `Pkg` standard library's documentation. If a new package's name does not adhere to those guidelines or could otherwise cause confusion, maintainers of the General registry or the Julia community might comment on the pull request to ask that the name be changed or why the usual conventions were not used in this case. 

Reviewer: ```suggestion
The guidelines for naming packages are listed [here](https://julialang.github.io/Pkg.jl/v1/creating-packages/#Package-naming-guidelines-1) in the package manager's documentation. If a new package's name does not adhere to those guidelines or could otherwise cause confusion, maintainers of the General registry or the Julia community might comment on the pull request to ask that the name be changed, or why the usual conventions were not used in this case.
